pkg/exec/exec.go

```diff
@@ -101,19 +101,19 @@ func RunWithStdoutReader(cmd Cmd, readerFunc func(io.Reader) error) error {
 		return err
 	}
 	defer pw.Close()
-	defer pr.Close()
 	cmd.SetStdout(pw)
 
 	errChan := make(chan error, 1)
 	go func() {
+		defer pr.Close()
 		errChan <- readerFunc(pr)
 	}()
 
 	err = cmd.Run()
-	err2 := <-errChan
 	if err != nil {
 		return err
 	}
+	err2 := <-errChan
```

Comment:
@neolit123 @BenTheElder I think that this was the deadlock, if the tar command failed it was waiting for the readerFunc to finish

Reply:
i would leave it to Ben to comment on the concurrency of exec.

Reply:
The reader func should not hang indefinitely, we can fix that instead.

Reply:
I thought that the reader func only finish when the pipe is closed

Reply:
er actually right, the pipe will be closed and this will exit with just _this_ change. the other changes are unnecessary

Reply:
but the other changes are cool 😎 , aren´t they?

Reply:
maybe? https://github.com/kubernetes-sigs/kind/pull/606/files#r293113238 should not happen

Reply:
what do you, mean the compress one or the concurrent? I´ve split the changes  in different commits so we can get rid of all of one of them